W-19804084-rtf-3.0-documentation-kt (#1592)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/deploy-mule-app-anypoint-controller.adoc

= Deploying and Managing Mule Applications Using Anypoint Controller for Kubernetes

Anypoint Controller is a Kubernetes operator that enables you to deploy and manage Mule applications on Runtime Fabric using GitOps workflows and Custom Resources (CR).

> Key Benefits:

Reviewer: ```suggestion
With Anypoint Controller, you can:
```

* Define application configurations declaratively in version control and let the controller automatically reconcile them with Anypoint Platform and your cluster.
* Use standard Kubernetes tools and practices to deploy Mule applications, including `kubectl`, Helm, and custom resources.
* Leverage Kubernetes features such as external secret managers, volume mounts, and RBAC policies to meet enterprise security requirements.

> This guide shows you how to:

Reviewer: Avoid the self referential "This guide".

Suggest using this list as a high-level process overview (numbered list with links to the docs where it's covered). See this section for an example: https://docs.mulesoft.com/api-governance/get-started#govern-your-apis 

Check out the overview guidelines, too: https://confluence.internal.salesforce.com/spaces/MTDT/pages/458629222/Writing+Product+Overviews

* Install and configure Anypoint Controller in your Runtime Fabric cluster.

> * Deploy Mule applications using declarative MuleApplication Custom Resources.

Reviewer: ```suggestion
* Deploy Mule applications using declarative Mule application Custom Resources.
```

Reviewer: (Not sure if that's correct.)

* Monitor application status using Kubernetes-native status conditions.
* Update, delete, and manage application lifecycles with standard Kubernetes commands.
* Integrate with CI/CD and GitOps tools like ArgoCD, FluxCD, and Terraform.
* Configure Kubernetes-specific settings including volumes, secrets, and node affinity.
* Troubleshoot common deployment and configuration issues.

== Scope and Limitations

[NOTE]
This feature is currently in Limited General Availability (Limited GA). To enable Anypoint Controller for your organization, contact MuleSoft Support.

> Supported features:

Reviewer: Are these features? Or are these prerequisites or compatibility?

* Runtime Fabric Mule application deployments
* Mule runtime engine (Mule) versions 4.6 and higher

> * Helm based Runtime Fabric installation or upgrade

Reviewer: ```suggestion
* Helm-based Runtime Fabric installation or upgrade
```

> Not supported features:

Reviewer: Should these be Limitations?

* CloudHub 2.0 Mule application deployments
* Scheduler-based Mule applications
* API proxy deployments

> * `rtfctl` based Runtime Fabric installation or upgrade

Reviewer: ```suggestion
* `rtfctl`-based Runtime Fabric installation or upgrade
```

[IMPORTANT]

> Backup and restore functionality (during Limited GA phase) is not supported. Plan your recovery strategy accordingly, leveraging your Git repository as the source of truth for application configurations.

Reviewer: ```suggestion
Backup and restore functionality (during the Limited GA phase) is not supported. Plan your recovery strategy accordingly, leveraging your Git repository as the source of truth for application configurations.
```

== Before You Begin

Before you begin, review these prerequisites and environment setup requirements.

> === Installation Method

Reviewer: ```suggestion
=== Installation Prerequisites
```

> To use Anypoint Controller, you must:

Reviewer: ```suggestion
To use Anypoint Controller, you must either:
```

> * Have an existing Runtime Fabric installation using Helm, or

Reviewer: ```suggestion
* Have an existing Runtime Fabric installation using Helm.
```

Reviewer: (Avoid this list structure.)

> * Install or upgrade Runtime Fabric using Helm (see xref:install-helm.adoc[Install Runtime Fabric Using Helm]) or xref:upgrade-helm.adoc[Upgrade Runtime Fabric Using Helm]

Reviewer: ```suggestion
* xref:install-helm.adoc[Install Runtime Fabric]) or xref:upgrade-helm.adoc[upgrade Runtime Fabric]  using Helm.
```

> === Management Constraints

Reviewer: ```suggestion
=== Management Limitations
```

Understanding the management model is critical before deploying and managing applications with Anypoint Controller:

> * Applications created via Mule Application Custom Resources can only be managed through the Custom Resource itself.

Reviewer: ```suggestion
* Applications created via Mule application Custom Resources can only be managed through the Custom Resource itself.
```

agent:
  rtfd:
    anypointController:
      enabled: true
----
[start=2]
. Upgrade the Runtime Fabric agent:

[source,bash]
----
helm upgrade rtf-agent <chart-repo>/rtf-agent \
  -f values.yaml \
  -n <rtf-namespace>
----
====

=== Define MuleApplication Custom Resource

Use the Custom Resource Definition (CRD) to define and manage Mule application configurations similar to any Kubernetes resources through Kubernetes APIs. Anypoint Controller watches the Mule Application Custom Resource (MuleApplication CR) state and matches it against the desired configured state.

Define your Mule application as a Custom Resource using your preferred deployment tool (such as YAML manifests with `kubectl`, Terraform, Helm charts, or GitOps tools). 

The Custom Resource specifies deployment configurations such as CPU and memory resources, replica count, runtime version, application artifact details, properties, volume mounts for secrets and configuration files, and Kubernetes-specific settings. 

After you apply this file to your Kubernetes cluster, for example by using `kubectl apply -f <filename>.yaml`, Anypoint Controller deploys the application based on the specifications.

> For a complete reference of all available MuleApplicationCR fields and configuration options, see xref:deploy-mule-app-anypoint-controller.adoc#understanding-custom-resource-definitions[Understanding Custom Resource Definitions].

Reviewer 2: ```suggestion
For a complete reference of all available MuleApplication CR fields and configuration options, see xref:deploy-mule-app-anypoint-controller.adoc#understanding-custom-resource-definitions[Understanding Custom Resource Definitions].
```

* `volumes` +
Defines the storage volumes available to the application, such as Kubernetes Secrets, ConfigMaps, persistent volumes, or external secret stores.
* `volumeMounts` +
Specifies where to mount the volumes inside the application container and makes the data accessible at the specified file path.

When you specify `volumes` and `volumeMounts` in the MuleApplication CR:

. The controller includes these specifications in the Deployment resource.
. Kubernetes mounts the volumes into the Mule application container at the specified paths.
. Your Mule application can access the files at runtime.

[NOTE]
Volumes specified in the MuleApplication CR are mounted to the main Mule application container, not to sidecar containers.

[%collapsible]
.Example: Mount Secrets from Kubernetes Secrets
====
Use Kubernetes native Secrets to store secret data on the Kubernetes cluster and mount them into your Mule application.

. Create a Secret in your Kubernetes cluster:

> 

Reviewer 2: ```suggestion
+
```

[source,bash]
----
kubectl create secret generic my-database-secret \
  --from-literal=db.username=admin \
  --from-literal=db.password=secret123 \
  -n my-apps
----
[start=2]

> .Mount it in your MuleApplication CR:

Reviewer 2: ```suggestion
. Mount it in your MuleApplication CR:
```

> 

Reviewer 2: ```suggestion
+
```

        sizeLimit: 1Gi
  volumeMounts:
    - name: temp-storage
      mountPath: /tmp/processing
----
====

[[create-an-anypoint-connected-app]]
=== Create a Connected App

Anypoint Controller requires a Connected App to authenticate with Anypoint Platform APIs. The Connected App acts on its own behalf using client credentials.

. Log in to Anypoint Platform.
. Navigate to *Access Management* > *Connected Apps*.
. Click *Create app*.
. Select *App acts on its own behalf (client credentials)*.
. Enter a name for your Connected App, such as `anypoint-controller`.
. Under *Scopes*, grant these permissions:
+
[%header%autowidth.spread]
|===
| Category | Permission | Description
| Exchange | Exchange Viewer | View and download application artifacts from Exchange
| OpenID | Profile | Access basic profile information
| Runtime Manager | Create Applications | Create new application deployments
| Runtime Manager | Delete Applications | Delete existing application deployments
| Runtime Manager | Read Runtime Fabrics | View Runtime Fabric target information
|===

[start=7]
. Click *Save*.
. Copy the *Client ID* and *Client Secret*. You need these values in the next step.

[IMPORTANT]

> Store the Client Secret securely. You cannot retrieve it again after leaving the create connected app page.

Reviewer 2: ```suggestion
Store the Client Secret securely. You can't retrieve it again after leaving the create connected app page.
```

=== Configure Anypoint Credentials

Create a Kubernetes Secret `anypoint-credentials` containing the Connected App credentials in the namespace where Runtime Fabric is installed.

[source,bash]
----
kubectl create secret generic anypoint-credentials \
  --from-literal=ANYPOINT_CLIENT_ID=<your-client-id> \
  --from-literal=ANYPOINT_CLIENT_SECRET=<your-client-secret> \
  -n <rtf-namespace>
----

> Replace the following:

Reviewer 2: ```suggestion
Replace these:
```

The number of replicas that are running and ready to serve traffic.

[source,yaml]
----
status:
  readyReplicas: 2
----

*Retry Information*

Indicates whether a failed operation can be automatically retried. When `true`, the controller retries the operation with exponential backoff.

[source,yaml]
----
status:
  retryable: true
----

*Reconciliation Metadata*

Used internally by the controller for change detection and reconciliation.

[source,yaml]
----
status:
  lastPublishedHash: "a1b2c3d4..."     # Hash of last published spec
  lastPublishedAt: "2024-01-15T10:30:00Z"  # Last publish timestamp
----
====

== Manage Mule Applications

Anypoint Controller enables you to manage the complete lifecycle of your Mule applications using standard Kubernetes commands and GitOps workflows.

=== Reconciliation Behavior

Anypoint Controller continuously reconciles the desired state:

* The controller synchronizes the application state every 30 seconds.
* The controller compares the CR spec hash with the last published hash to detect changes.
* When changes are detected, the controller updates Anypoint Platform automatically.
* Failed operations are retried with exponential backoff.

> * The controller periodically refreshes state even without changes (every 10 minutes).

Reviewer 3: ```suggestion
* Synchronizes the application state every 30 seconds.
* Compares the CR spec hash with the last published hash to detect changes.
* When changes are detected, updates Anypoint Platform automatically.
* Retries failed operations with exponential backoff.
* Refreshes the state every 10 minutes, even when no changes occur.
```

Manage your application lifecycle:

* <<update-a-mule-application,Update a Mule Application>>

> * <<delete-a-muleapplication,Delete a MuleApplication>>

Reviewer 3: ```suggestion
* <<delete-a-mule-application,Delete a Mule Application>>
```

> === Update a Mule Application

Reviewer 3: ```suggestion
[[update-a-mule-application]]
=== Update a Mule Application
```

. Modify your Mule application YAML file:

[source,yaml]
----
spec:
  replicas: 5  # Changed from 2 to 5
----
[start=2]
. Apply the change:

[source,bash]
----
kubectl apply -f hello-world-app.yaml
----
[start=3]
. Monitor the update:

[source,bash]
----
kubectl get muleapplication hello-world-app -n my-apps -w
----

The controller detects the change and initiates the update automatically.

[NOTE]
Some fields like `target`, `businessGroup`, and `environment` are immutable and cannot be changed after creation. For a complete list of mutable and immutable fields, see xref:deploy-mule-app-anypoint-controller.adoc#understanding-custom-resource-definitions[Understanding Custom Resource Definitions].

> === Delete a MuleApplication

Reviewer 3: ```suggestion
[[delete-a-mule-application]]
=== Delete a Mule Application
```

> To delete an application managed by Anypoint Controller, delete the MuleApplication CR:

Reviewer 3: ```suggestion
To delete an application managed by Anypoint Controller, delete the `muleapplication` CR:
```

[source,bash]
----
kubectl delete muleapplication <app-name> -n <namespace>
----

Example:

[source,bash]
----
kubectl delete muleapplication hello-world-app -n my-apps
----

The controller automatically:

. Deletes the application from Anypoint Platform.

> . Removes all Kubernetes resources (Deployment, Service, ConfigMaps, Secrets).

Reviewer 3: ```suggestion
. Removes all Kubernetes resources (Deployment, Service, ConfigMaps, and Secrets).
```

. Removes the finalizer from the CR.
. Allows Kubernetes to delete the CR.

[IMPORTANT]

> Deletion is permanent. Ensure your version control repository contains the application configuration if you need to redeploy later.

Reviewer 3: ```suggestion
Deletion is permanent. Make sure your version control repository contains the application configuration in case you need to redeploy it later.
```

== Integrate with CI/CD and GitOps Tools

Anypoint Controller works seamlessly with GitOps and Infrastructure-as-Code tools.

=== GitOps Workflow

. Store MuleApplication manifests in a version control repository.
. Define branch strategies (for example, `dev`, `staging`, `production`).
. Use pull requests or merge requests for configuration changes.
. Automate deployment with GitOps tools.

Integrate Anypoint Controller with these tools:

* <<argocd-integration,ArgoCD Integration>>
* <<fluxcd-integration,FluxCD Integration>>
* <<terraform-integration,Terraform Integration>>

=== ArgoCD Integration

. Store your manifests in version control:
+
[source,console]
----
my-mule-apps/
├── production/
│   ├── hello-world-app.yaml
│   └── order-processing-app.yaml
└── staging/
    ├── hello-world-app.yaml
    └── order-processing-app.yaml
----
[start=2]
. Create an ArgoCD Application:

> 

Reviewer 2: ```suggestion
+
```

[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: mule-apps-production
  namespace: argocd
spec:
  project: default
  source:
    repoURL: https://github.com/myorg/mule-deployments
    targetRevision: main
    path: production
  destination:
    server: https://kubernetes.default.svc
    namespace: mule-apps
  syncPolicy:
    automated:
      prune: true       # Delete resources removed from version control
      selfHeal: true    # Revert manual changes
    syncOptions:
      - CreateNamespace=true
----

ArgoCD continuously monitors the version control repository and automatically deploys changes.

=== FluxCD Integration

. Create a GitRepository resource:

> 

Reviewer 2: ```suggestion
+
```

[source,yaml]
----
apiVersion: source.toolkit.fluxcd.io/v1
kind: GitRepository
metadata:
  name: mule-apps
  namespace: flux-system
spec:
  interval: 1m
  url: https://github.com/myorg/mule-deployments
  ref:
    branch: main
----
[start=2]
. Create a Kustomization resource:

> 

Reviewer 2: ```suggestion
+
```

==== Migrating from Templates

If you currently use KubernetesTemplate for configuration, migrate those settings to the MuleApplication CR fields:

[%header%autowidth.spread]
|===
| Template Field | MuleApplication CR Field
| Pod labels | `spec.pod.metadata.labels`
| Pod annotations | `spec.pod.metadata.annotations`
| Node selectors | `spec.pod.nodeSelector`
| Tolerations | `spec.pod.tolerations`
| Affinity rules | `spec.pod.affinity`
| Topology spread | `spec.pod.topologySpreadConstraints`
| Service annotations | `spec.service.metadata.annotations`
| Deployment labels | `spec.deployment.metadata.labels`
|===

== Troubleshooting

Use these techniques to diagnose and resolve common issues with Anypoint Controller and Mule application deployments.

=== Check Controller Logs

Anypoint Controller is a feature enabled in the existing `rtfd` container. View the logs to troubleshoot deployment issues:

[source,bash]
----
kubectl logs -l app=rtf-agent -n <rtf-namespace> -c anypoint-controller --tail=100
----

Follow logs in real-time:

[source,bash]
----
kubectl logs -l app=rtf-agent -n <rtf-namespace> -c anypoint-controller -f
----

Enable Debug Logs:

> By default, the log level is set to `info`, which does not emit detailed information for diagnosis. To enable debug logs for more detailed troubleshooting, add the following environment variables to the `app-config` ConfigMap and restart the agent deployment:

Reviewer 2: ```suggestion
By default, the log level is set to `info`, which does not emit detailed information for diagnosis. To enable debug logs for more detailed troubleshooting, add these environment variables to the `app-config` ConfigMap and restart the agent deployment:
```

----

=== Common Issues

[%collapsible]
.Application Not Reconciling
====
The `Applied` condition remains `False` or changes aren't reflected.

To resolve this issue:

. Verify the namespace has the correct label:
+
[source,bash]
----
kubectl get namespace <namespace> --show-labels
----
+
The namespace should have `rtf.mulesoft.com/agentNamespace=<rtf-namespace>`.
+
Add the label if missing:
+
[source,bash]
----
kubectl label namespace <namespace> rtf.mulesoft.com/agentNamespace=<rtf-namespace>
----

. Check the Connected App credentials:
+
[source,bash]
----
kubectl get secret anypoint-credentials -n <rtf-namespace> -o yaml
----

. Verify the Connected App has required permissions in Anypoint Platform.
. Check <<check-controller-logs>> for errors.
====

[%collapsible]
.Application Deployment Fails
====

The `Applied` condition is `True`, but `Available` is `False`.

To resolve this issue:

. Check pod status:
+
[source,bash]
----
kubectl get pods -n <namespace> -l rtf.mulesoft.com/app=<app-name>
----

. View pod logs:
+
[source,bash]
----
kubectl logs <pod-name> -n <namespace>
----

. Check pod events:
+
[source,bash]
----
kubectl describe pod <pod-name> -n <namespace>
----

> . Common causes:

Reviewer 2: maybe move this to before the steps?

   - Invalid runtime version
   - Artifact not found in Exchange
   - Insufficient resources (CPU/memory)
   - Image pull errors
====

[%collapsible]
.Invalid Configuration
====

`kubectl apply` fails with validation errors.

To resolve this issue:

. Validate the YAML syntax:
+
[source,bash]
----
kubectl apply -f app.yaml --dry-run=server
----

> . Check immutable field violations (target, businessGroup, environment cannot be changed after creation).

Reviewer 2: ```suggestion
. Check immutable field violations (target, businessGroup, and environment can't be changed after creation).
```

. Verify required fields are present.
====

[%collapsible]
.Anypoint Platform API Errors
====
The `Applied` condition shows `APIErrorResponse`.

To resolve this issue:

. Read the error message in the status condition.
+

>  Common errors:

Reviewer 2: this is rendering as a code block

   - `409 Conflict`: Application name already exists (created outside Anypoint Controller)
   - `404 Not Found`: Invalid target, business group, or environment name
   - `403 Forbidden`: Connected App lacks permissions
   - `400 Bad Request`: Invalid configuration (for example, runtime version)

> . Verify configuration matches your Anypoint Platform setup.

Reviewer 2: ```suggestion
. Verify the configuration matches your Anypoint Platform setup.
```
